### DISCUSSION TITLE
Open rewrite LST and mass refactoring

### DISCUSSION THREAD

**Discussion by gshenoy123** [openrewrite/rewrite#3602] [General] [answered]
Hello all, 

We would like few pointers please on Open rewrite's _mass refactoring topic_. 

In the documentation, there are following statements - 

- Document reference - https://docs.openrewrite.org/concepts-explanations/lossless-semantic-trees

> Since OpenRewrite LSTs are serializable, we can store them as a byproduct of compilation and then operate on them en masse later.
> Once we have a serialized LST for a particular source file, and since it also contains type information, it can be refactored/searched completely independently of other source files in the same source package or repository. This makes mass search and refactoring an approximately linearly scalable operation.

We use open rewrite recipes to apply refactoring per project, but as the document mentions _LST can be stored then operate on them en masse later_. It would be helpful to understand the statements above on mass refactoring aspect - does these mean that each project will be iterated over and then resultant LST would be stored and then recipes applied on these LSTs, if so, how does mass refactoring with serializable LSTs work and how do recipes get applied?

**Comment by timtebeek** [comment, marked as answer]
Hi @gshenoy123 ; thanks for the thorough read of the documentation, and posting your question here. The mass refactoring aspect is something we offer through Moderne, which you can use either through https://app.moderne.io or [the Moderne CLI](https://docs.moderne.io/moderne-cli/cli-intro) you can download from there.

The mass refactoring is unlocked through serialization, which is a proprietary feature of Moderne, not part of the Open Source OpenRewrite plugins. That said we do encourage use for Open Source projects through the Moderne platform, and allow limited CLI use for projects on your own machine as well.

The [Moderne Architecture documentation](https://docs.moderne.io/architecture-readme/architecture) goes more in depth [how LST artifacts are produced](https://docs.moderne.io/references/how-ast-artifacts-are-produced) and used to run recipes and create pull requests.

I hope that helps answers your question; if it raises any more feel free to ask those here, or reach out through https://getmoderne.com/

**Reply by timtebeek** [reply]
We've since updated the documentation to be more clear of what's available through OpenRewrite, and what Moderne additionally offers. Thanks for bringing this to our attention. I hope you agree that resolves this question.

**Reply by gshenoy123** [reply]
Hi @timtebeek, thank you for your reply. This is helpful and answers our query.

**Reply by shenoy-ganesh-kumar** [reply]
Hi @timtebeek,

Couple of queries more below, can you please let us know -

1. What is the performance improvement comparing OSS open rewrite and Moderne CLI when it comes to executing recipes. We understand once code is compiled, we can continuously apply multiple recipes on generated LSTs. Can you please let us know on any benchmarks if you have conducted.
2. From your reply we understand CLI is proprietary and hence would have commercial license and the usage here you mentioned is as trial version. Can we run a standalone version in our environment without connectivity to Moderne portal.

**Reply by timtebeek** [reply]
Hi @shenoy-ganesh-kumar ! I'll reply to each of your questions below.

> What is the performance improvement comparing OSS open rewrite and Moderne CLI when it comes to executing recipes. We understand once code is compiled, we can continuously apply multiple recipes on generated LSTs. Can you please let us know on any benchmarks if you have conducted.

The OpenRewrite OSS plugins for Maven and Gradle allow you to update one project at a time, and require you to build an LST before each recipe run. This can add up to a lot of time depending on the number and size of projects. The Moderne CLI and Moderne Platform allow you to run against multiple projects at the same time, using pre-built LSTs. These also allow you to run against _large_ applications where the LST would otherwise not fit into memory, by using serialization. As such the use cases are different, and we therefor do not run benchmarks for a direct comparison.

> From your reply we understand CLI is proprietary and hence would have commercial license and the usage here you mentioned is as trial version. Can we run a standalone version in our environment without connectivity to Moderne portal.

You can absolutely run a standalone instance without a connection to Moderne. For that we have developed the Moderne Developer Experience, which we'll introduce with a blog some time this week or next. If you want to learn more I recommend you fill out https://getmoderne.com/, as that allows us to start a private conversation about your specifics.

Hope that helps!

**Comment by gshenoy123** [comment]
Hi @timtebeek, Thank you, yes this is helpful. We have noted the points you mentioned regarding developer experience, and also we will connect through the medium you mentioned on further queries.

**Reply by timtebeek** [reply]
Glad to hear! Hope we can follow up in private to help you out. :)

Meanwhile, here's a blog we posted yesterday [about the Moderne CLI](https://www.moderne.io/blog/moderne-cli-multi-repo-developer-experience-automated-code-refactoring) that might be interesting to read already.
